README.md
Desafio: Consulta Processual
===

Oii, tudo bem? Seja bem vindo a Busca de Processos. 

A linguagem usada neste projeto é Javascript com framework React.js e Next, os testes são com Jest(unitários) e Cypress (e2e).

## 1. Iniciando o projeto 

Para começar, rode:

`npm install`

`npm run dev`

*Irá rodar no caminho: `http://localhost:3000/`*

Para rodar os testes unitários do front-end, rode em um segundo terminal:

`npm run test`


Para rodar o testes e2e cypress, você pode rodar o comando no segundo terminal ou criar um terceiro :

`npx cypress run`
   

## 2. Overview

A arquitetura do projeto possui 4 pastas principais:

1. components: onde fica os componentes do projeto para serem reutilizados 
2. mock: onde fica os dados mockados e utilizados pelo projeto, que seriam os Tribunais (courtMock.json) e os dados das movimentações (processesMock.json)
3. pages: onde renderiza e reutiliza os componentes criados, por exemplo a pasta process que contém a tela de exibição
4. store: responsável por consumir os dados dos mocks e compartilhar dados entre components (guardar estado)


Na home o usuário tem a opção de buscar por número de CNJ de um processo e por um tribunal especifico do CNJ.
O usuário pode também refazer a operação de busca em qualquer outra página a partir do Header da página.

Testes unitários são realizados nos componentes e nas páginas para verificar o uso ideial deles e renderização na tela. O fluxo de sucesso e erros é testado a partir do teste cypress e2e que se encontra no caminho `cypress\e2e\spec.cy.js`


## 3. Mocks

Existem alguns dados mockados dentro de `processMock.json`, localizado em `src/mock`, eles deverão ser utilizados como meio de busca, sendo assim, irei disponibilizar todos os ID's disponíveis para facilitar o uso da aplicação:

Processo n° 1 = CNJ: 1508252-80.2023.4.04.0058 | Tribunal: TJSP

Processo n° 2 = CNJ: 2968371-16.2023.2.00.1120 | Tribunal: TJMG

Processo n° 3 = CNJ: 8065227-04.2023.5.18.2672 | Tribunal: TJRS

Processo n° 4 = CNJ: 7596542-92.2023.5.22.8364 | Tribunal: TJSC

Processo n° 5 = CNJ: 1607604-38.2023.5.14.3715 | Tribunal: TJPR

## 4. Decisões 

Se tiver alguma dúvida sobre as decisões por favor acesse o arquivo mais detalhado:
`HISTORY.MD`
[https://github.com/daniellemadrid/Consulta-Processual/blob/main/HISTORY.MD]


HISTORY.MD
Reflexões e decisões
===


## Frontend

Eu pensei primeiro em como iria montar a estrutura, me veio em mente a estrutura Atomic Design mas logo pensei em algo mais simples já que o projeto não seria tão complexo. Então acabei criando componentes com arquivos de testes juntos em cada pasta. 
Construí com Next pois é uma ferramenta nova que me possibilita ter mais praticidade em desenvolver, por exemplo: com ele consegui criar rotas a partir de um arquivo ou pasta dentro de src/pages. (tudo o que for criado dentro de pages é interpretado como um endpoint), caso seja criado um arquivo chamado jusBrasil dentro de src/pages, será possível acessar direto no navegador.

Criei os componentes e chamei para duas páginas criadas, src/index para buscar e src/pages/process para exibir, me baseei pelas imagens fornecidas pelo repositório e criei validações para o input de texto ser obrigatório e enviar o cnj de forma válida junto com o tribunal correspondente.

Usei Tailwind para estilizar o projeto pois estou aprendendo e pareceu uma ótima oportunidade para aplicar. Os ganhos com a utilização do mesmo são notáveis, os quais são: praticidade em poder manter em um arquivo só, ou seja, não é necessário a criação de arquivos CSS para cada página e/ou componente.

## Backend

No backend é utilizado Mocks e Zustand para a criação de Stores. 
Zustand é uma biblioteca open source que neste projeto teve a finalidade de compartilhar dados entre components e pages, assim como, um useContext. A diferença entre Zustand e um useContext está na escalabilidade e praticidade presente no Zustand, dentro da store de process.js é possível localizar o que corresponderia a um GET convencional. Lá dentro tem o equivalente a um findUnique de um banco de dados comum que retorna os dados selecionados com base em sua pesquisa, porém, como está sendo utilizado Mocks, foi feito find dentro da Store de process para consumir o Mock de processos e verificar se os dados que o usuário informou estão corretos e devolver o processo que corresponde ao cnj e tj inseridos. Esses dados são enviados para a página de process (src/pages/process/index) responsável por renderizar as movimentaçãoes e dados dos processos.

Pensando em aproximar a solução utilizando Mocks em algo mais realista, foi criado dois arquivos, um deles sendo responsável por armazenar os dados dos tribunais e o outro responsável por guardar os processos, cada tribunal tem um id e cada processo carrega o id de um tribunal. Sendo assim, a simulação de uma relação entre tabelas Many To One (muitos processos para um tribunal).


## Testes e outros

Para testes unitários eu usei Jest e Testing-Library, então cada componente tem seu respectivo teste para verificar a renderização na tela e as páginas, criei testes para verificar se as validações que citei estavam funcionando. 
Usei o Cypress para testar o fluxo do projeto pensando nas validações também. Não testei muito sobre os componentes da tela com Cypress porque o Jest já está fazendo isso. 

## Ideias 
Se eu pudesse, persistiria os dados com banco de dados real para substituir os Mocks criados e com isso conseguir utilizar 100% do potencial do Next, já que ele é um framework feito para facilitar a construção de projetos onde o frontend e backend se encontram no mesmo repositório, encurtando vários processos. Passaria para Atomic Design pensando mais a longo prazo e escalabilidade do projeto, pensaria em novos cenários de testes para ter uma maior garantia de cobertura e qualidade do projeto.

## Bugs conhecidos

Estou ciente do bug que ocorre na página home onde é necessário clicar duas vezes no botão de busca para realizar a ação, este item seria prioridade caso houvesse mais tempo.

src/pages/home/home.test.js
import React from 'react';
import { render, screen } from '@testing-library/react';
import Home from '.';
import '@testing-library/jest-dom'

jest.mock('next/router', () => require('next-router-mock'));

describe('Home page', () => {

    it('should render the home page', () => {
        render(<Home />)

        expect(screen.getByText("Busca de processos")).toBeInTheDocument();
        //   expect(screen.getByText("Selecione um tribunal para listar os processos ou busque pelo número unificado de um processo específico.")).toBeInTheDocument();

        const dropdownElement = screen.getByText('Tribunais');
        expect(dropdownElement).toBeInTheDocument();

        const searchBarElement = screen.getByRole('textbox');
        expect(searchBarElement).toBeInTheDocument();
        expect(screen.getByTestId("button")).toBeInTheDocument();

    });
});

src/pages/process/index.test.js
import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom'
import Process from './index.jsx';


jest.mock('next/router', () => require('next-router-mock'));

describe('Process page', () => {

    it('should render the Process page', () => {
        render(<Process />)

        const buttonBackElement = screen.getByTestId('buttonBack');
        expect(buttonBackElement).toBeInTheDocument();

        expect(screen.getByTestId("button")).toBeInTheDocument();

        const searchBarElement = screen.getByRole('textbox');
        expect(searchBarElement).toBeInTheDocument();

        const dropdownElement = screen.getByText('Tribunais');
        expect(dropdownElement).toBeInTheDocument();

        const numberOfProcessElement = screen.getByText('Processo n. do');
        expect(numberOfProcessElement).toBeInTheDocument();

        const dateOfProcessElement = screen.getByText('Distribuído em:');
        expect(dateOfProcessElement).toBeInTheDocument();

        const movementsElement = screen.getByText('Movimentações');
        expect(movementsElement).toBeInTheDocument();

        const detailsElement = screen.getByText('Detalhes do processo');
        expect(detailsElement).toBeInTheDocument();

        const courtElement = screen.getByText('Endereço do tribunal');
        expect(courtElement).toBeInTheDocument();

        const processDetailsElement = screen.getByText('Partes envolvidas');
        expect(processDetailsElement).toBeInTheDocument();

        const authorElement = screen.getByText('Parte envolvida - Exequente');
        expect(authorElement).toBeInTheDocument();

        const defendedElement = screen.getByText('Parte envolvida - Apelante');
        expect(defendedElement).toBeInTheDocument();

    });

});

tailwind.config.js
/** @type {import('tailwindcss').Config} */
module.exports = {
    content: [
        './src/pug/*.pug',
        './src/html/*.html',
        "./src/**/*.{js,ts,jsx,tsx}",
    ],
    theme: {
        screens: {
            sm: '640px',
            md: '768px',
            lg: '1024px',
            xl: '1156px',
            '2xl': '1450px'
        },
        spacing: {
            px: '1px',
            px: '1px',
            '0': '0px',
            '0.5': '0.125rem',
            '1': '0.25rem',
            '1.5': '0.375rem',
            '2': '0.5rem',
            '2.5': '0.625rem',
            '3': '0.75rem',
            '3.5': '0.875rem',
            '4': '1rem',
            '5': '1.25rem',
            '6': '1.5rem',
            '7': '1.75rem',
            '8': '2rem',
            '9': '2.25rem',
            '10': '2.5rem',
            '11': '2.75rem',
            '12': '3rem',
            '14': '3.5rem',
            '16': '4rem',
            '20': '5rem',
            '24': '6rem',
            '28': '7rem',
            '32': '8rem',
            '36': '9rem',
            '40': '10rem',
            '44': '11rem',
            '48': '12rem',
            '52': '13rem',
            '56': '14rem',
            '60': '15rem',
            '64': '16rem',
            '72': '18rem',
            '80': '20rem',
            '96': '24rem',
            '112': '28rem',
            '128': '32rem',
            '144': '36rem',
        },
        backgroundColor: theme => ({
            ...theme('colors'),
            body: '#F1F5FB',
        }),
        backgroundImage: {
            none: 'none',
            'gradient-to-t': 'linear-gradient(to top, var(--tw-gradient-stops))',
            'gradient-to-tr': 'linear-gradient(to top right, var(--tw-gradient-stops))',
            'gradient-to-r': 'linear-gradient(to right, var(--tw-gradient-stops))',
            'gradient-to-br': 'linear-gradient(to bottom right, var(--tw-gradient-stops))',
            'gradient-to-b': 'linear-gradient(to bottom, var(--tw-gradient-stops))',
            'gradient-to-bl': 'linear-gradient(to bottom left, var(--tw-gradient-stops))',
            'gradient-to-l': 'linear-gradient(to left, var(--tw-gradient-stops))',
            'gradient-to-tl': 'linear-gradient(to top left, var(--tw-gradient-stops))',
        },
        backgroundPosition: {
            bottom: 'bottom',
            center: 'center',
            left: 'left',
            'left-bottom': 'left bottom',
            'left-top': 'left top',
            right: 'right',
            'right-bottom': 'right bottom',
            'right-top': 'right top',
            top: 'top',
        },
        backgroundSize: {
            auto: 'auto',
            cover: 'cover',
            contain: 'contain',
        },
        borderColor: theme => ({
            ...theme('colors'),
            DEFAULT: '#E1E4E8',
        }),
        borderRadius: {
            none: '0',
            sm: '0.125rem',
            DEFAULT: '0.25rem',
            md: '0.375rem',
            lg: '0.5rem',
            xl: '0.75rem',
            '2xl': '1rem',
            '3xl': '1.5rem',
            full: '9999px',
        },
        borderWidth: {
            DEFAULT: '1px',
            '0': '0',
            '2': '2px',
            '4': '4px',
            '8': '8px',
        },
        boxShadow: {
            sm: '0 1px 2px 0 rgba(0, 0, 0, 0.05)',
            DEFAULT: '0px 4px 8px -4px #15181C14',
            md: '0 4px 6px -1px rgba(0, 0, 0, 0.1), 0 2px 4px -1px rgba(0, 0, 0, 0.06)',
            lg: '0 10px 15px -3px rgba(0, 0, 0, 0.1), 0 4px 6px -2px rgba(0, 0, 0, 0.05)',
            xl: '0 20px 25px -5px rgba(0, 0, 0, 0.1), 0 10px 10px -5px rgba(0, 0, 0, 0.04)',
            '2xl': '0 25px 50px -12px rgba(0, 0, 0, 0.25)',
            inner: 'inset 0 2px 4px 0 rgba(0, 0, 0, 0.06)',
            none: 'none',
        },
        container: {},
        cursor: {
            auto: 'auto',
            DEFAULT: 'default',
            pointer: 'pointer',
            wait: 'wait',
            text: 'text',
            move: 'move',
            'not-allowed': 'not-allowed',
        },
        flex: {
            '1': '1 1 0%',
            auto: '1 1 auto',
            initial: '0 1 auto',
            none: 'none',
        },
        flexGrow: {
            '0': '0',
            DEFAULT: '1',
        },
        flexShrink: {
            '0': '0',
            DEFAULT: '1',
        },
        fontFamily: {
            body: '"DM Sans", ui-sans-serif, system-ui, -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, "Noto Sans", sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol", "Noto Color Emoji"',
            heading: '"DM Sans", ui-sans-serif, system-ui, -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, "Noto Sans", sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol", "Noto Color Emoji"',
            sans: '"DM Sans", ui-sans-serif, system-ui, -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, "Noto Sans", sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol", "Noto Color Emoji"',
            serif: 'ui-serif, Georgia, Cambria, "Times New Roman", Times, serif',
            mono: 'ui-monospace, SFMono-Regular, Menlo, Monaco, Consolas, "Liberation Mono", "Courier New", monospace',
        },
        fontSize: {
            xs: ['0.75rem', { lineHeight: '1rem' }],
            sm: ['0.875rem', { lineHeight: '1.25rem' }],
            base: ['1rem', { lineHeight: '1.5rem' }],
            lg: ['1.125rem', { lineHeight: '1.75rem' }],
            xl: ['1.25rem', { lineHeight: '1.75rem' }],
            '2xl': ['1.5rem', { lineHeight: '2rem' }],
            '3xl': ['1.875rem', { lineHeight: '2.25rem' }],
            '4xl': ['2.25rem', { lineHeight: '2.5rem' }],
            '5xl': ['3rem', { lineHeight: '1.16' }],
            '6xl': ['3.75rem', { lineHeight: '1' }],
            '7xl': ['4.5rem', { lineHeight: '1' }],
            '8xl': ['6rem', { lineHeight: '1' }],
            '9xl': ['8rem', { lineHeight: '1' }],
        },
        fontWeight: {
            thin: '100',
            extralight: '200',
            light: '300',
            normal: '400',
            medium: '500',
            semibold: '600',
            bold: '700',
            extrabold: '800',
            black: '900',
        },
        height: theme => ({
            auto: 'auto',
            ...theme('spacing'),
            full: '100%',
            screen: '100vh',
        }),
        inset: (theme, { negative }) => ({
            auto: 'auto',
            ...theme('spacing'),
            ...negative(theme('spacing')),
            '1/2': '50%',
            '1/3': '33.333333%',
            '2/3': '66.666667%',
            '1/4': '25%',
            '2/4': '50%',
            '3/4': '75%',
            full: '100%',
            '-1/2': '-50%',
            '-1/3': '-33.333333%',
            '-2/3': '-66.666667%',
            '-1/4': '-25%',
            '-2/4': '-50%',
            '-3/4': '-75%',
            '-full': '-100%',
        }),
        letterSpacing: {
            tighter: '-0.05em',
            tight: '-0.025em',
            normal: '0',
            wide: '0.025em',
            wider: '0.05em',
            widest: '0.1em',
        },
        lineHeight: {
            none: '1',
            tight: '1.25',
            snug: '1.375',
            normal: '1.5',
            relaxed: '1.625',
            loose: '2',
        },
        listStyleType: {
            none: 'none',
            disc: 'disc',
            decimal: 'decimal',
        },
        margin: (theme, { negative }) => ({
            auto: 'auto',
            ...theme('spacing'),
            ...negative(theme('spacing')),
        }),
        maxHeight: {
            full: '100%',
            screen: '100vh',
            none: 'none',
            xs: '20rem',
            sm: '24rem',
            md: '28rem',
            lg: '32rem',
            xl: '36rem',
            '2xl': '42rem',
            '3xl': '48rem',
            '4xl': '56rem',
            '5xl': '64rem',
            '6xl': '72rem',
            '7xl': '80rem',
        },
        maxWidth: {
            none: 'none',
            xs: '20rem',
            sm: '24rem',
            md: '28rem',
            lg: '32rem',
            xl: '36rem',
            '2xl': '42rem',
            '3xl': '48rem',
            '4xl': '56rem',
            '5xl': '64rem',
            '6xl': '72rem',
            '7xl': '80rem',
            full: '100%',
            min: 'min-content',
            fit: 'fit-content',
            max: 'max-content',
            prose: '65ch',
        },
        minHeight: {
            '0': '0',
            full: '100%',
            screen: '100vh',
        },
        minWidth: {
            '0': '0',
            full: '100%',
            fit: 'fit-content',
        },
        objectPosition: {
            bottom: 'bottom',
            center: 'center',
            left: 'left',
            'left-bottom': 'left bottom',
            'left-top': 'left top',
            right: 'right',
            'right-bottom': 'right bottom',
            'right-top': 'right top',
            top: 'top',
        },
        order: {
            first: '-9999',
            last: '9999',
            none: '0',
            '1': '1',
            '2': '2',
            '3': '3',
            '4': '4',
            '5': '5',
            '6': '6',
            '7': '7',
            '8': '8',
            '9': '9',
            '10': '10',
            '11': '11',
            '12': '12',
        },
        padding: theme => theme('spacing'),
        placeholderColor: theme => theme('colors'),
        stroke: {
            current: 'currentColor',
        },
        textColor: theme => ({
            ...theme('colors'),
            body: '#15181C',
        }),
        width: theme => ({
            auto: 'auto',
            ...theme('spacing'),
            '1/2': '50%',
            '1/3': '33.333333%',
            '2/3': '66.666667%',
            '1/4': '25%',
            '2/4': '50%',
            '3/4': '75%',
            '1/5': '20%',
            '2/5': '40%',
            '3/5': '60%',
            '4/5': '80%',
            '1/6': '16.666667%',
            '2/6': '33.333333%',
            '3/6': '50%',
            '4/6': '66.666667%',
            '5/6': '83.333333%',
            '1/12': '8.333333%',
            '2/12': '16.666667%',
            '3/12': '25%',
            '4/12': '33.333333%',
            '5/12': '41.666667%',
            '6/12': '50%',
            '7/12': '58.333333%',
            '8/12': '66.666667%',
            '9/12': '75%',
            '10/12': '83.333333%',
            '11/12': '91.666667%',
            'fit': 'fit-content',
            full: '100%',
            screen: '100vw',
        }),
        zIndex: {
            auto: 'auto',
            '0': '0',
            '10': '10',
            '20': '20',
            '30': '30',
            '40': '40',
            '50': '50',
        },
        extend: {
            colors: {
                indigo: {
                    50: '#EBEEFD',
                    100: '#D8DEFB',
                    200: '#B1BDF8',
                    300: '#8A9CF4',
                    400: '#637BF1',
                    500: '#3D5AEE',
                    600: '#2D23B1',
                    700: '#221A85',
                    800: '#161258',
                    900: '#0B092C',
                }
            }
        }
    },
    corePlugins: {},
}

src/store/processes.js
import { create } from "zustand"
const processesMock = require("../mock/processesMock.json")

const mappedProcesses = processesMock.map((process) => process)
const regex = /^\d{7}-\d{2}\.\d{4}\.\d{1}\.\d{2}\.\d{4}$/gm

export const useProcessStore = create((set, get) => ({
    processes: [],
    processId: "",
    tjId: "",
    tjName: "",
    errorMessage: "",
    process: [],

    setProcesses: (processes) => set({ processes }),
    setErrorMessage: (errorMessage) => set({ errorMessage }),
    setTjId: (tjId) => set({ tjId }),
    setTjName: (tjName) => set({ tjName }),
    setProcessId: (processId) => set({ processId }),
    clearData: () => set({ processId: null, tjId: null, tjName: null, errorMessage: null, process: null }),
    findProcess: async (processId, tjId) => {
        set({ errorMessage: '' })
        if (processId == null || processId == '') {
            set({ errorMessage: "Digite um número de processo" })
            return
        }
        if (!tjId) {
            set({ errorMessage: "Selecione um tribunal" })
            return
        }
        if (mappedProcesses.find((process) => process.id == processId && process.idTJ != tjId)) {
            set({ errorMessage: "Processo não existe no tribunal selecionado" })
            return
        }
        if (mappedProcesses.find((process) => process.id == processId) == null) {
            set({ errorMessage: "Número do processo não encontrado" })
            return
        }
        if (get().processId == processId) {
            return
        }
        if (get().errorMessage == '') {
            set({ process: mappedProcesses.find((process) => process.id === processId) })
        }
    }
}))

cypress/e2e/spec.cy.js
describe('home page test', () => {
  beforeEach(() => {
    cy.visit('http://localhost:3000/home');
  });
  it('sucess case', () => {
    cy.contains('Busca de processos').should('be.visible')
    cy.get('[data-testid="input"]').type('2968371-16.2023.2.00.1120');
    cy.contains('Tribunais').click()
    cy.contains('TJMG').click()
    cy.get('[data-testid="button"]').click()
    cy.get('[data-testid="button"]').click()
    cy.url().should('eq', 'http://localhost:3000/process')
    cy.get('[data-testid="processNumber"]').should('be.visible');
    cy.contains('Distribuído em:').should('be.visible');
    cy.contains('Movimentações').should('be.visible');
    cy.contains('Detalhes do processo').should('be.visible');
    cy.contains('Endereço do tribunal').should('be.visible');
    cy.contains('Valor a ser pago').should('be.visible');
    cy.contains('Partes envolvidas').should('be.visible');

  });

  it('error when trying to send empty input ', () => {
    cy.get('[data-testid="button"]').click()
    cy.contains('Digite um número de processo');

  });

  it('error when trying to send empty dropdown', () => {
    cy.get('[data-testid="input"]').type('2968371-16.2023.2.00.1120');
    cy.get('[data-testid="button"]').click()
    cy.contains('Selecione um tribunal');

  });

  it('error sending process with wrong court ', () => {
    cy.get('[data-testid="input"]').type('2968371-16.2023.2.00.1120');
    cy.contains('Tribunais').click()
    cy.contains('TJSP').click()
    cy.get('[data-testid="button"]').click()
    cy.contains('Processo não existe no tribunal selecionado');

  });


  it('shows error when process is not found with cnj', () => {
    cy.get('[data-testid="input"]').type('22344');
    cy.contains('Tribunais').click()
    cy.contains('TJPR').click()
    cy.get('[data-testid="button"]').click()
    cy.contains(
      'Número do processo não encontrado'
    );
  });
})



